source/base/hdf5.cc

```diff
@@ -0,0 +1,1513 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2017 - 2018 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#include <deal.II/base/config.h>
+
+#ifdef DEAL_II_WITH_HDF5
+
+#  include <deal.II/base/hdf5.h>
+
+#  include <deal.II/lac/full_matrix.h>
+#  include <deal.II/lac/vector.h>
+
+#  include <hdf5.h>
+
+#  include <memory>
+#  include <numeric>
+#  include <vector>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace HDF5
+
+{
+  namespace internal
+  {
+    // This function gives the HDF5 datatype corresponding to the C++ type. In
+    // the case of std::complex types the HDF5 handlers are automatically freed
+    // using the destructor of std::shared_ptr.
+    template <typename number>
+    std::shared_ptr<hid_t>
```

Comment:
I added the following comment:

In the case of `std::complex` types the HDF5 handlers are automatically freed using the destructor of `std::shared_ptr`. `std::shared_ptr` is used instead of `std::unique_ptr` because the destructor of `std::shared_ptr` doesn't have to be defined in the template argument. In the other hand, the destructor of `std::unique` has to be defined in the template argument. Native types such as H5T_NATIVE_DOUBLE does not require a destructor. but compound types such as `std::complex<double>` require a destructor to free the HDF5 resources.

Reply:
I see. I suspect that the performance impact for using `std::shared_ptr` are negligible.